File: problems/27-segment-tree/003-range-min-max-query/java/Solution.java
public class Solution {
    private int[] minTree, maxTree;
    private int n;

    public void build(int[] a) {
        n = a.length;
        minTree = new int[4 * n];
        maxTree = new int[4 * n];
        build(1, 0, n - 1, a);
    }

    private void build(int node, int l, int r, int[] a) {
        if (l == r) { minTree[node] = maxTree[node] = a[l]; return; }
        int m = l + (r - l) / 2;
        build(node * 2, l, m, a); build(node * 2 + 1, m + 1, r, a);
        pull(node);
    }

    private void pull(int node) {
        minTree[node] = Math.min(minTree[node * 2], minTree[node * 2 + 1]);
        maxTree[node] = Math.max(maxTree[node * 2], maxTree[node * 2 + 1]);
    }

    public int[] query(int ql, int qr) {
        int[] result = query(1, 0, n - 1, ql, qr);
        return result;
    }

    private int[] query(int node, int l, int r, int ql, int qr) {
        if (qr < l || r < ql) return new int[]{Integer.MAX_VALUE, Integer.MIN_VALUE};
        if (ql <= l && r <= qr) return new int[]{minTree[node], maxTree[node]};
        int m = l + (r - l) / 2;
        int[] a = query(node * 2, l, m, ql, qr);
        int[] b = query(node * 2 + 1, m + 1, r, ql, qr);
        return new int[]{Math.min(a[0], b[0]), Math.max(a[1], b[1])};
    }

    public void update(int index, int value) { update(1, 0, n - 1, index, value); }

    private void update(int node, int l, int r, int index, int value) {
        if (l == r) { minTree[node] = maxTree[node] = value; return; }
        int m = l + (r - l) / 2;
        if (index <= m) update(node * 2, l, m, index, value);
        else update(node * 2 + 1, m + 1, r, index, value);
        pull(node);
    }
}

File: problems/27-segment-tree/003-range-min-max-query/python/solution.py
class SegmentTree:
    def __init__(self, nums):
        self.n = len(nums)
        self.min_tree = [0] * (4 * self.n)
        self.max_tree = [0] * (4 * self.n)
        self._build(1, 0, self.n - 1, nums)

    def _build(self, node, left, right, nums):
        if left == right:
            self.min_tree[node] = self.max_tree[node] = nums[left]
            return
        mid = (left + right) // 2
        self._build(node * 2, left, mid, nums)
        self._build(node * 2 + 1, mid + 1, right, nums)
        self._pull(node)

    def _pull(self, node):
        self.min_tree[node] = min(self.min_tree[node * 2], self.min_tree[node * 2 + 1])
        self.max_tree[node] = max(self.max_tree[node * 2], self.max_tree[node * 2 + 1])

    def query(self, ql, qr):
        return self._query(1, 0, self.n - 1, ql, qr)

    def _query(self, node, left, right, ql, qr):
        if qr < left or right < ql:
            return float('inf'), float('-inf')
        if ql <= left and right <= qr:
            return self.min_tree[node], self.max_tree[node]
        mid = (left + right) // 2
        a = self._query(node * 2, left, mid, ql, qr)
        b = self._query(node * 2 + 1, mid + 1, right, ql, qr)
        return min(a[0], b[0]), max(a[1], b[1])

    def update(self, index, value):
        self._update(1, 0, self.n - 1, index, value)

    def _update(self, node, left, right, index, value):
        if left == right:
            self.min_tree[node] = self.max_tree[node] = value
            return
        mid = (left + right) // 2
        if index <= mid:
            self._update(node * 2, left, mid, index, value)
        else:
            self._update(node * 2 + 1, mid + 1, right, index, value)
        self._pull(node)
